### DISCUSSION TITLE
Let neural network output potential instead of density?

### DISCUSSION THREAD

**Discussion by gaowutong** [darioizzo/geodesyNets#110] [Ideas]
Hello, 

First, I must say it's an elegant idea! The density field gives a direct insight into the internal mass distributions of small bodies.

In your paper, the acceleration is computed by an integration. Therefore, the number of density samples should be carefully selected to balance the computation time and accuracy. I'm wondering if letting the network output potential instead of density can solve this issue. Moreover, by taking the first and second derivatives of the potential w.r.t. the position, we can easily get the acceleration and density, since differentiating a neural network is less expensive than integrating it.

I'm not sure this works, could you please give some advice about that?

**Comment by gomezzz** [comment]
> Hello, 
> 
> First, I must say it's an elegant idea! The density field gives a direct insight into the internal mass distributions of small bodies.
> 
> In your paper, the acceleration is computed by an integration. Therefore, the number of density samples should be carefully selected to balance the computation time and accuracy. I'm wondering if letting the network output potential instead of density can solve this issue. Moreover, by taking the first and second derivatives of the potential w.r.t. the position, we can easily get the acceleration and density, since differentiating a neural network is less expensive than integrating it.
> 
> I'm not sure this works, could you please give some advice about that?

Hey,

thanks for your kind words. In general, we are not 100% sure which factors limit the approach the most. Computational time wasn't really a problem, accuracy of the integration could be. But there are other factors (such as quality of training data, different network architectures etc.)

On training for the potential, this would, if I understand correctly, mean that we learn the gravitational field instead of the density distribution. Depending on your application this may actually introduce other problem because you may want to reconstruct shape and density from that which may be tricky too?

If the only goal is to match the expected accelerations, then it could be beneficial though I guess? :thinking: 

Not sure what @darioizzo thinks :)

**Reply by gaowutong** [reply]
Thanks for your reply!

To reconstruct the potential, it may possible recover the shape and density, because the gravitational potential $V$ satisfies poisson equation inside the body, i.e., $\Delta{V(x,y,z)} = -4\pi G \rho$.

This means by taking the laplacian operator of the potential (which can be implemented by automated differentiation in deep learning), we can compute the density. This may be less expensive and more accurate than integtration.

**Comment by darioizzo** [comment]
If the network represents the potential you would have to learn the model parameters from potential readings in points x,y,z outside the asteroid. Then the Poisson equation would need to be evaluated in points inside the asteroid to get the density.  In other words you would be constantly using points outside the training set, and even worse their derivatives! That seems to be detrimental to the expected accuracy.

This is a reason I liked the network to represent directly the density! It allows to learn from points outside the body information inside the body, forcing implicitly to fulfill the Poisson equation.

That said, I did not try what you suggested and it may be worth investigating.
